actors/market/src/lib.rs

```diff
@@ -530,93 +530,147 @@ impl Actor {
         Ok(VerifyDealsForActivationReturn { sectors: sectors_data })
     }
 
-    /// Activate a set of deals, returning the combined deal space and extra info for verified deals.
-    fn activate_deals(
+    /// Activate a set of deals, returning the deal space and extra info for sectors containing
+    /// verified deals. Sectors are activated in parameter-defined order and can fail independently of
+    /// each other with the responsible ExitCode recorded in a BatchReturn.
+    fn batch_activate_deals(
         rt: &impl Runtime,
-        params: ActivateDealsParams,
-    ) -> Result<ActivateDealsResult, ActorError> {
+        params: BatchActivateDealsParams,
+    ) -> Result<BatchActivateDealsResult, ActorError> {
         rt.validate_immediate_caller_type(std::iter::once(&Type::Miner))?;
         let miner_addr = rt.message().caller();
         let curr_epoch = rt.curr_epoch();
 
-        let (deal_spaces, verified_infos) = rt.transaction(|st: &mut State, rt| {
-            let proposal_array = st.get_proposal_array(rt.store())?;
-            let proposals = get_proposals(&proposal_array, &params.deal_ids, st.next_id)?;
+        let (activations, batch_ret) = rt.transaction(|st: &mut State, rt| {
+            let mut deal_states: Vec<(DealID, DealState)> = vec![];
+            let mut batch_gen = BatchReturnGen::new(params.sectors.len());
+            let mut activations: Vec<DealActivation> = vec![];
+            let mut activated_deals: HashSet<DealID> = HashSet::new();
+
+            for p in params.sectors {
+                let proposal_array = st.get_proposal_array(rt.store())?;
```

Comment:
Unfortunately I didn't see a way to move this outside of the loop due to a reborrow of st when calling `remove_pending_deal_allocation_id`